code/game/machinery/computer/prisoner.dm

```diff
@@ -24,118 +21,149 @@
 	if(..())
-		return 1
-
-	usr.set_machine(src)
-
-	if(href_list["id"])
-		switch(href_list["id"])
-			if("0")
-				var/obj/item/card/id/prisoner/I = usr.get_active_hand()
-				if(istype(I))
-					usr.drop_item()
-					I.loc = src
-					inserted_id = I
-				else
-					to_chat(usr, "<span class='warning'>No valid ID.</span>")
-			if("1")
-				inserted_id.loc = get_step(src,get_turf(usr))
+		return
+
+	add_fingerprint(usr)
+
+	if(ui_act_modal(action, params, ui))
+		return
+	if(ui_login_act(action, params))
+		return
+
+	var/mob/living/user = ui.user
+
+
+
+	switch(action)
+		if("id_card")
+			if(inserted_id)
+				inserted_id.forceMove(loc)
 				inserted_id = null
-			if("2")
-				inserted_id.mining_points = 0
-			if("3")
-				var/num = round(input(usr, "Choose prisoner's goal:", "Input an Integer", null) as num|null)
-				if(num >= 0)
-					inserted_id.goal = num
-	if(href_list["inject1"])
-		var/obj/item/implant/I = locate(href_list["inject1"])
-		if(I)	I.activate(1)
-
-	else if(href_list["inject5"])
-		var/obj/item/implant/I = locate(href_list["inject5"])
-		if(I)	I.activate(5)
-
-	else if(href_list["inject10"])
-		var/obj/item/implant/I = locate(href_list["inject10"])
-		if(I)	I.activate(10)
-
-	else if(href_list["lock"])
-		if(src.allowed(usr))
-			authenticated = !authenticated
+				return
+			var/obj/item/card/id/prisoner/I = user.get_active_hand()
+			if(istype(I) && user.drop_item())
+				I.forceMove(src)
+				inserted_id = I
+			else
+				to_chat(user, "<span class='warning'>No valid ID.</span>")
+		if("inject")
+			var/obj/item/implant/chem/implant = locateUID(params["uid"])
+			implant.activate(text2num(params["amount"]))
+		if("reset_points")
+			if(inserted_id)
+				inserted_id.points = 0
+
+
+/obj/machinery/computer/prisoner/proc/ui_act_modal(action, list/params, datum/tgui/ui)
+	if(!ui_login_get().logged_in)
+		return
+	. = TRUE
+	var/id = params["id"]
+	var/mob/living/user = ui.user
+	var/list/arguments = istext(params["arguments"]) ? json_decode(params["arguments"]) : params["arguments"]
+	switch(ui_modal_act(src, action, params))
+		if(UI_MODAL_OPEN)
+			switch(id)
+				if("warn")
+					ui_modal_input(src, id, "Please enter your message:", null, arguments = list(
+						"uid" = arguments["uid"],
+					))
+				if("set_points")
+					ui_modal_input(src, id, "Please enter the new point goal:", null, arguments)
+				else
+					return FALSE
+		if(UI_MODAL_ANSWER)
+			var/answer = params["answer"]
+			switch(id)
+				if("warn")
+					var/obj/item/implant/tracking/implant = locateUID(arguments["uid"])
+					if(!implant)
+						return
+					if(implant.warn_cooldown >= world.time)
+						to_chat(user, "<span class='warning'>The warning system for that implant is still cooling down.</span>")
+						return
+					implant.warn_cooldown = world.time + IMPLANT_WARN_COOLDOWN
+					if(implant.imp_in)
+						var/mob/living/carbon/implantee = implant.imp_in
+						var/warning = copytext(sanitize(text2num(answer)), 1, MAX_NAME_LEN)
```

Comment:
50 characters max? Should be plenty for most cases. But it might be nice to give feedback to the user if it goes over 50. Or give them a warning prior to sending the message that the max length is 50.
`MAX_MESSAGE_LEN` can otherwise be used which equates to 1024

Reply:
done